ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Tables/UtxoDiff.hs

```diff
@@ -0,0 +1,177 @@
+{-# LANGUAGE DeriveAnyClass             #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingVia                #-}
+{-# LANGUAGE InstanceSigs               #-}
+
+module Ouroboros.Consensus.Ledger.Tables.UtxoDiff (
+    -- * Types
+    UtxoDiff (..)
+    -- * Conversion
+  , keysSet
+    -- * Construction
+  , diff
+    -- ** Maps
+  , fromMapDeletes
+  , fromMapInserts
+  , fromListDeletes
+  , fromListInserts
+    -- * Query
+    -- ** Size
+  , null
+  , numDeletes
+  , numInserts
+  , size
+    -- * Applying diffs
+  , applyUtxoDiff
+  , applyUtxoDiffForKeys
+    -- * Filter
+  , filterOnlyKey
+  ) where
+
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Monoid
+import           Data.Semigroup.Cancellative
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           GHC.Generics
+import           NoThunks.Class
+import           Prelude hiding (null)
+
+{------------------------------------------------------------------------------
+  Types
+------------------------------------------------------------------------------}
+
+data UtxoDiff k v = UtxoDiff (Map k v) (Map k v)
+  deriving stock (Show, Eq)
+  deriving Generic
+  deriving NoThunks
+
+instance Functor (UtxoDiff k) where
+  fmap f (UtxoDiff m1 m2) = UtxoDiff (Map.map f m1) (Map.map f m2)
+
+instance Ord k => Semigroup (UtxoDiff k v) where
+  -- right biased
+  (UtxoDiff m1 m2) <> (UtxoDiff m3 m4) =
+     UtxoDiff
+        -- all deletes except the ones cancelling m2
+        (Map.union m1 (m3 `Map.difference` m2))
+        -- inserts from the first component without the latest deletes and inserts
+        -- from the second component
+        (Map.unionWith const (m2 `Map.difference` m3) m4)
+
+instance Ord k => Monoid (UtxoDiff k v) where
+  mempty :: UtxoDiff k v
+  mempty = UtxoDiff mempty mempty
+
+{------------------------------------------------------------------------------
+  Conversion
+------------------------------------------------------------------------------}
+
+keysSet :: Ord k => UtxoDiff k v -> Set k
+keysSet (UtxoDiff m1 m2) = Map.keysSet m1 <> Map.keysSet m2
+
+{------------------------------------------------------------------------------
+  Construction
+------------------------------------------------------------------------------}
+
+diff :: Map k v -> Map k v -> UtxoDiff k v
+diff = UtxoDiff
```

Comment:
Why does this works? Seems counterintuitive. Should it not be something like:

```suggestion
diff :: Map k v -> Map k v -> UtxoDiff k v
diff m1 m2 = UtxoDiff (m1 `Map.difference` m2) (m2 `Map.difference` m1)
```

If it's just the application of the `UtxoDiff` constructor that you need, then I'd suggest exposing that